Content.Server/GameObjects/Components/Atmos/GasTankComponent.cs

```diff
@@ -1,10 +1,323 @@
+            OpenInterface(actor.playerSession);
+            return true;
+        }
+
+        public void Activate(ActivateEventArgs eventArgs)
+        {
+            if (!eventArgs.User.TryGetComponent(out IActorComponent? actor)) return;
+            OpenInterface(actor.playerSession);
+        }
+
+        public void ConnectToInternals()
+        {
+            if (IsConnected || !IsFunctional) return;
+            var internals = GetInternalsComponent();
+            if (internals == null) return;
+            IsConnected = internals.TryConnectTank(Owner);
+            UpdateUserInterface();
+        }
+
+        public void DisconnectFromInternals(IEntity? owner = null)
+        {
+            if (!IsConnected) return;
+            IsConnected = false;
+            GetInternalsComponent(owner)?.DisconnectTank();
+            UpdateUserInterface();
+        }
+
+        private void UpdateUserInterface(bool initialUpdate = false)
+        {
+            _userInterface?.SetState(
+                new GasTankBoundUserInterfaceState
+                {
+                    TankPressure = Air?.Pressure ?? 0,
+                    OutputPressure = initialUpdate ? OutputPressure : (float?) null,
+                    InternalsConnected = IsConnected,
+                    CanConnectInternals = IsFunctional && GetInternalsComponent() != null
+                });
+        }
+
+        private void UserInterfaceOnOnReceiveMessage(ServerBoundUserInterfaceMessage message)
+        {
+            switch (message.Message)
+            {
+                case GasTankSetPressureMessage msg:
+                    OutputPressure = msg.Pressure;
+                    break;
+                case GasTankToggleInternalsMessage _:
+                    ToggleInternals();
+                    break;
+            }
+        }
+
+        private void ToggleInternals()
+        {
+            if (IsConnected)
+            {
+                DisconnectFromInternals();
+                return;
+            }
+
+            ConnectToInternals();
+        }
+
+        private InternalsComponent? GetInternalsComponent(IEntity? owner = null)
+        {
+            if (owner != null) return owner.GetComponentOrNull<InternalsComponent>();
+            return ContainerHelpers.TryGetContainer(Owner, out var container)
+                ? container.Owner.GetComponentOrNull<InternalsComponent>()
+                : null;
+        }
+
+        public void AssumeAir(GasMixture giver)
+        {
+            Air?.Merge(giver);
+            CheckStatus();
+        }
+
+        private void CheckStatus()
+        {
+            if (Air == null)
+                return;
+
+            var pressure = Air.Pressure;
+
+            if (pressure > Atmospherics.TankFragmentPressure)
+            {
+                // Give the gas a chance to build up more pressure.
+                Air.React(this);
+                Air.React(this);
+                Air.React(this);
```

Comment:
Though this does look funny I feel like you might as well put it in fori loop.

Reply:
sad, fixed